### DISCUSSION TITLE
Want to contribute?

### DISCUSSION THREAD

**Discussion by Derpius** [TAServers/issues#16] [Announcements]
Whether you an experienced developer with years of experience, or a beginner looking to go beyond E2 and Starfall, we’re always welcoming of contributions.

To help you help us, I’ve compiled below a list of existing and potential TAS projects (mostly for the GMod server) which could suit you.

If starting a new project, we can help bootstrap you with a proper GitHub repository setup, the best tooling for the project, as well as general project management and technical guidance.

❗️There’s no requirement for your work to be part of a TAS organisation repo, but it makes it easier to collaborate.

# Ongoing Projects

These are in active development and are either ready or almost ready for additional collaborators.

- [GMod Custom Weapons](https://github.com/TAServers/sweps)
  - Addon for small custom weapons that are too small to be their own addon
  - An easy repo to contribute too, as you don't need to learn the existing codebase
- Lest
  - Lua testing framework written in Lua, but also designed to work with NPM
  - Ready for dev with a backlog of work scoped out
- New relay
  - Next-gen scalable chat relay
  - Not quite ready yet, needs full test coverage and CI for acceptable DevEx

In addition to the above, we also have projects like buildmode which don’t see active development, but can always be improved if you have any ideas.

# New Project Ideas

Here’s some ideas for projects you could start, ordered by difficulty. As said above, we can help you get started on the right track.

# Beginner

- Custom stools
  - Same as the custom sweps, we can create an stools monorepo for these
- Discord leave notifications
  - Small Discord bot/webhook (in your language of choice) to send a message to the new members channel when a user leaves the server
  - The messages should ideally match the system join messages as close as possible (you can omit the random puns)

# Intermediate

- Bespoke HUD
  - Quite a design focused one, with a lot of complexity coming from needing to work with Awesomium (if you use HTML)
  - Would also need to do a bit of user research
- Live E2/Starfall editor display
  - An opt-in display that appears above/adjacent to a player showing the code they’re editing in real time

# Advanced

- E2 and Starfall live collaboration
  - Quite a technically complex one for improving collaboration
  - This would involve synchronising changes and managing conflicts in real time between at least 2 collaborators
  - It should ideally be decoupled from any specific editor/addon as much as possible (i.e., the syncing can be separated from e2 and sf)
- GMod polling system (if you're thinking of picking this up, talk to me to get a full set of requirements)
  - A GMod addon which allows admins to create persistent polls for players on the server
  - This will let us take votes on gmod-specific topics, such as changing the default map, without excluding any part of our player base
  - Users should be notified that there are X polls they haven't voted on when joining the server
    - This could be a chat message
  - The admin who created a poll should be able to update, close and reopen the poll at any time
  - There should be a UI that allows users to view current and past polls, and allows admins to create new or update their existing polls

# Afterword

If you have any other ideas for projects you want to work on then let us know. Or if you just want to suggest something for others to work on, then you can post a suggestion in the comments here or as a new discussion.

**Comment by yogwoggf** [comment]
Ok cool i will help. YOu guys. Good server :) Like it lots .

**Reply by ELPatitoLoco** [reply]
:)
